File: src/main/java/eu/codetopic/utils/ui/activity/navigation/NavigationActivity.java
package eu.codetopic.utils.ui.activity.navigation;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.PersistableBundle;
import android.support.annotation.ColorInt;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.internal.NavigationMenuPresenter;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.internal.view.SupportMenuItem;
import android.support.v4.view.GravityCompat;
import android.support.v4.view.ViewCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.view.menu.MenuItemImpl;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.lang.reflect.Field;

import eu.codetopic.java.utils.Objects;
import eu.codetopic.java.utils.log.Log;
import eu.codetopic.utils.AndroidUtils;
import eu.codetopic.utils.R;
import eu.codetopic.utils.ui.activity.fragment.BaseFragmentActivity;

public abstract class NavigationActivity extends BaseFragmentActivity {

    private static final String LOG_TAG = "NavigationActivity";

    private static final String KEY_SWITCHING_ACCOUNTS =
            "eu.codetopic.utils.ui.activity.navigation.NavigationActivity.SWITCHING_ACCOUNTS";

    private boolean mEnableSwitchingAccounts = false;
    private boolean mSwitchingAccounts = false;

    private DrawerLayout mDrawer;
    private NavigationView mNavigationView;
    private LinearLayout mHeaderView;

    private ImageView mAppIconView;

    private FrameLayout mContainerAppTitle;
    private TextView mTextViewAppTitle;

    private LinearLayout mContainerAccountsSwitch;
    private TextView mTextViewAccountName;
    private ImageButton mImageButtonEditAccount;
    private ImageView mImageViewAccountChangeArrow;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState != null) {
            mSwitchingAccounts = savedInstanceState.getBoolean(KEY_SWITCHING_ACCOUNTS, false);
        }

        setContentView(R.layout.navigation_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        mDrawer = findViewById(R.id.drawer_layout);
        mNavigationView = findViewById(R.id.nav_view);
        mHeaderView = mNavigationView.getHeaderView(0).findViewById(R.id.navigationHeader);

        mAppIconView = mHeaderView.findViewById(R.id.imageViewAppIcon);

        mContainerAppTitle = mHeaderView.findViewById(R.id.container_app_title);
        mTextViewAppTitle = mContainerAppTitle.findViewById(R.id.textViewAppTitle);

        mContainerAccountsSwitch = mHeaderView.findViewById(R.id.container_accounts_switch);
        mContainerAccountsSwitch.setOnClickListener(v -> toggleSwitchingAccounts());
        mTextViewAccountName = mContainerAccountsSwitch.findViewById(R.id.textViewAccountName);
        mImageButtonEditAccount = mContainerAccountsSwitch.findViewById(R.id.imageButtonEditAccount);
        mImageButtonEditAccount.setOnClickListener(this::onEditAccountButtonClick);
        mImageViewAccountChangeArrow = mContainerAccountsSwitch.findViewById(R.id.imageViewAccountChangeArrow);

        setNavigationViewAppIconDrawable(AndroidUtils.getActivityIcon(this, getComponentName()));
        setHeaderBackgroundColor(AndroidUtils.getColorFromAttr(this,
                R.attr.colorPrimaryDark, Color.rgb(0, 0, 0)));

        mTextViewAppTitle.setText(AndroidUtils.getAppLabel(this));
        setSupportActionBar(toolbar);

        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(this, mDrawer, toolbar,
                R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        mDrawer.addDrawerListener(toggle);
        toggle.syncState();

        mNavigationView.setNavigationItemSelectedListener(item -> {
                    mDrawer.closeDrawer(GravityCompat.START);
                    if (mEnableSwitchingAccounts && mSwitchingAccounts) {
                        setSwitchingAccounts(false);
                        return onAccountNavigationItemSelected(item);
                    }
                    return onNavigationItemSelected(item);
                }
        );
        invalidateNavigationMenu();
    }

    protected Class<? extends Fragment> getMainFragmentClass() {
        return null;
    }

    protected Fragment getMainFragment() {
        try {
            return getMainFragmentClass().newInstance();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    protected Fragment onCreateMainFragment() {
        try {
            return getMainFragment();
        } catch (Exception e) {
            Log.e(LOG_TAG, "onCreateMainFragment - provided wrong MainFragment", e);
            return null;
        }
    }

    public void setActiveAccountEditButtonEnabled(boolean enabled) {
        mImageButtonEditAccount.setVisibility(enabled ? View.VISIBLE : View.GONE);
    }

    public boolean isActiveAccountEditButtonEnabled() {
        return mImageButtonEditAccount.getVisibility() == View.VISIBLE;
    }

    public void setEnableSwitchingAccounts(boolean enableSwitchingAccounts) {
        mEnableSwitchingAccounts = enableSwitchingAccounts;
        invalidateNavigationMenu();
    }

    public boolean isEnableSwitchingAccounts() {
        return mEnableSwitchingAccounts;
    }

    public boolean isSwitchingAccounts() {
        return mSwitchingAccounts;
    }

    public void setSwitchingAccounts(boolean switchingAccounts) {
        mSwitchingAccounts = switchingAccounts;
        invalidateNavigationMenu();
    }

    public boolean toggleSwitchingAccounts() {
        mSwitchingAccounts = !mSwitchingAccounts;
        invalidateNavigationMenu();
        return mSwitchingAccounts;
    }

    public void setNavigationViewAppIconBitmap(Bitmap bm) {
        mAppIconView.setImageBitmap(bm);
    }

    public void setNavigationViewAppIconDrawable(@Nullable Drawable drawable) {
        mAppIconView.setImageDrawable(drawable);
    }

    public void setNavigationViewAppIconResource(@DrawableRes int resId) {
        mAppIconView.setImageResource(resId);
    }

    public void setHeaderBackground(Drawable drawable) {
        ViewCompat.setBackground(mHeaderView, drawable);
    }

    public void setHeaderBackgroundColor(@ColorInt int color) {
        mHeaderView.setBackgroundColor(color);
    }

    public void setHeaderBackgroundResource(@DrawableRes int resId) {
        mHeaderView.setBackgroundResource(resId);
    }

    public void setNavigationTitle(CharSequence title) {
        mTextViewAppTitle.setText(title);
    }

    /*@Override
    public void setTitle(CharSequence title) {
        super.setTitle(title);
        if (mTextViewAppTitle != null) mTextViewAppTitle.setText(title);
    }*/

    @Override
    public void onSaveInstanceState(Bundle outState, PersistableBundle outPersistentState) {
        super.onSaveInstanceState(outState, outPersistentState);
        outState.putBoolean(KEY_SWITCHING_ACCOUNTS, mSwitchingAccounts);
    }

    @Override
    protected void onDestroy() {
        mDrawer = null;
        mNavigationView = null;
        mTextViewAppTitle = null;
        super.onDestroy();
    }

    @Override
    public void onBackPressed() {
        if (mDrawer.isDrawerOpen(GravityCompat.START)) {
            mDrawer.closeDrawer(GravityCompat.START);
            return;
        }

        Fragment fragment = getCurrentFragment();
        Class<? extends Fragment> mainFragmentClass = getMainFragmentClass();
        if (fragment == null || !Objects.equals(fragment.getClass(), mainFragmentClass)) {
            replaceFragment(mainFragmentClass);
            return;
        }

        super.onBackPressed();
    }

    @SuppressLint("RestrictedApi")
    public void invalidateNavigationMenu() {
        if (mNavigationView == null) return;

        if (mEnableSwitchingAccounts) {
            mTextViewAccountName.setText(onUpdateActiveAccountName());
            mImageViewAccountChangeArrow.setImageResource(mSwitchingAccounts
                    ? R.drawable.ic_arrow_drop_up
                    : R.drawable.ic_arrow_drop_down);
            mContainerAppTitle.setVisibility(View.GONE);
            mContainerAccountsSwitch.setVisibility(View.VISIBLE);
        } else {
            mContainerAccountsSwitch.setVisibility(View.GONE);
            mContainerAppTitle.setVisibility(View.VISIBLE);
        }

        NavigationMenuPresenter presenter = null;
        try {
            Field presenterField = NavigationView.class.getDeclaredField("mPresenter");
            presenterField.setAccessible(true);
            presenter = (NavigationMenuPresenter) presenterField.get(mNavigationView);//this is HACK
        } catch (Exception e) {
            Log.w(LOG_TAG, "invalidateNavigationMenu - " +
                    "can't get menu presenter: can't get field from class", e);
        }

        if (presenter != null) presenter.setUpdateSuspended(true);

        Menu menu = mNavigationView.getMenu();
        menu.clear();

        if (mEnableSwitchingAccounts && mSwitchingAccounts) {
            onCreateAccountNavigationMenu(menu);
            // setupMenuItemsClickListeners(menu);
        } else {
            onCreateNavigationMenu(menu);
            // setupMenuItemsClickListeners(menu);
        }
        resetNavigationView(getCurrentFragment());

        if (presenter != null) {
            presenter.setUpdateSuspended(false);
            presenter.updateMenuView(false);
        }
    }

    /*private void setupMenuItemsClickListeners(@NonNull Menu menu) {
        Field listenerField;
        try {
            listenerField = MenuItemImpl.class.getDeclaredField("mClickListener");
            listenerField.setAccessible(true);
        } catch (NoSuchFieldException e) {
            Log.e(LOG_TAG, "setupMenuItemsClickListeners - " +
                    "can't setup listeners: can't get field from class", e);
            return;
        }

        for (int i = 0, size = menu.size(); i < size; i++) {
            MenuItem item = menu.getItem(i);
            try {
                if (item instanceof MenuItemImpl) {
                    final SupportMenuItem.OnMenuItemClickListener listener =
                            (SupportMenuItem.OnMenuItemClickListener) listenerField.get(item);//this is HACK
                    listenerField.set(item, (SupportMenuItem.OnMenuItemClickListener) item1 -> {
                        boolean result = listener != null && listener.onMenuItemClick(item1);
                        mDrawer.closeDrawer(GravityCompat.START);
                        return result;
                    });
                } else throw new ClassCastException("Wrong class: " + item.getClass());
            } catch (Throwable e) {
                Log.e(LOG_TAG, "setupMenuItemsClickListeners - " +
                        "can't setup listener for " + item, e);
            }
            if (item.hasSubMenu()) setupMenuItemsClickListeners(item.getSubMenu());
        }
    }*/

    private void resetNavigationView(@Nullable Fragment currentFragment) {
        if (mNavigationView == null) return;
        if (mEnableSwitchingAccounts && mSwitchingAccounts) {
            onUpdateSelectedAccountNavigationMenuItem(currentFragment, mNavigationView.getMenu());
            mTextViewAccountName.setText(onUpdateActiveAccountName());
        } else {
            onUpdateSelectedNavigationMenuItem(currentFragment, mNavigationView.getMenu());
        }
    }

    protected boolean onCreateNavigationMenu(@NonNull Menu menu) {
        return false;
    }

    protected boolean onCreateAccountNavigationMenu(@NonNull Menu menu) {
        return false;
    }

    /**
     * Override this method to enable navigation menu items selecting
     *
     * @param currentFragment current fragment
     * @param menu            navigation menu
     * @return true if checked state of any item was changed
     */
    protected boolean onUpdateSelectedNavigationMenuItem(@Nullable Fragment currentFragment, @NonNull Menu menu) {
        if (currentFragment != null) {
            Log.e(LOG_TAG, "onUpdateSelectedNavigationMenuItem can't detect selected item for " + currentFragment.getClass() +
                    ", override method onUpdateSelectedNavigationMenuItem() and implement your own selected item detecting.");
            return false;
        }


        for (int i = 0; i < menu.size(); i++) {
            menu.getItem(i).setChecked(false);
        }
        return true;
    }

    protected boolean onUpdateSelectedAccountNavigationMenuItem(@Nullable Fragment currentFragment, @NonNull Menu menu) {
        Log.e(LOG_TAG, "onUpdateSelectedAccountNavigationMenuItem can't update selected account item" +
                ", override method onUpdateSelectedAccountNavigationMenuItem() and implement selected item detecting.");
        return false;
    }

    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        return false;
    }

    public boolean onAccountNavigationItemSelected(@NonNull MenuItem item) {
        return false;
    }

    protected CharSequence onUpdateActiveAccountName() {
        Log.e(LOG_TAG, "onUpdateActiveAccountName can't get active account name" +
                ", override method onUpdateActiveAccountName() and implement active account name detecting.");
        return "";
    }

    protected boolean onEditAccountButtonClick(View v) {
        return false;
    }

    @Override
    public void onAttachFragment(Fragment fragment) {
        super.onAttachFragment(fragment);
        if (CURRENT_FRAGMENT_TAG.equals(fragment.getTag())) resetNavigationView(fragment);
    }

    @Override
    @SuppressWarnings("PrivateResource")
    protected void onBeforeReplaceFragment(FragmentTransaction ft, Fragment fragment) {
        ft.setCustomAnimations(R.anim.abc_fade_in, R.anim.abc_fade_out);
    }
}

File: src/main/java/eu/codetopic/utils/ui/activity/navigation/NavigationActivity.kt
/*
 * utils
 * Copyright (C)   2017  anty
 *
 * This program is free  software: you can redistribute it and/or modify
 * it under the terms  of the GNU General Public License as published by
 * the Free Software  Foundation, either version 3 of the License, or
 * (at your option) any  later version.
 *
 * This program is distributed in the hope that it  will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied  warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You  should have received a copy of the GNU General Public License
 * along  with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package eu.codetopic.utils.ui.activity.navigation

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.os.PersistableBundle
import android.support.annotation.ColorInt
import android.support.annotation.DrawableRes
import android.support.design.internal.NavigationMenuPresenter
import android.support.design.widget.NavigationView
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentTransaction
import android.support.v4.view.GravityCompat
import android.support.v4.view.ViewCompat
import android.support.v7.app.ActionBarDrawerToggle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import com.mikepenz.google_material_typeface_library.GoogleMaterial

import eu.codetopic.java.utils.log.Log
import eu.codetopic.utils.AndroidExtensions.getIconics
import eu.codetopic.utils.AndroidUtils
import eu.codetopic.utils.R
import eu.codetopic.utils.ui.activity.fragment.BaseFragmentActivity
import kotlinx.android.extensions.CacheImplementation
import kotlinx.android.extensions.ContainerOptions
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.activity_navigation_header.*
import kotlinx.android.synthetic.main.activity_navigation_header.view.*
import kotlinx.android.synthetic.main.activity_navigation_base.*
import kotlinx.android.synthetic.main.activity_module_toolbar.*

@ContainerOptions(CacheImplementation.SPARSE_ARRAY)
abstract class NavigationActivity : BaseFragmentActivity() {

    companion object {

        private const val LOG_TAG = "NavigationActivity"

        private const val KEY_SWITCHING_ACCOUNTS =
                "eu.codetopic.utils.ui.activity.navigation.NavigationActivity.SWITCHING_ACCOUNTS"
    }

    @ContainerOptions(CacheImplementation.SPARSE_ARRAY)
    internal inner class HeaderViews : LayoutContainer {
        override val containerView: LinearLayout =
                this@NavigationActivity.navigationView.getHeaderView(0).boxHeader
    }

    private lateinit var header: HeaderViews

    private lateinit var drawerToggle: ActionBarDrawerToggle

    var enableSwitchingAccounts = false
        set(enable) {
            field = enable
            invalidateNavigationMenu()
        }
    private var switchingAccounts = false
    var isSwitchingAccounts: Boolean
        get() = switchingAccounts
        set(switchingAccounts) {
            this.switchingAccounts = switchingAccounts
            invalidateNavigationMenu()
        }

    var enableActiveAccountEditButton: Boolean
        get() = header.butAccountEdit.visibility == View.VISIBLE
        set(enabled) {
            header.butAccountEdit.visibility =
                    if (enabled) View.VISIBLE
                    else View.GONE
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        savedInstanceState?.let {
            switchingAccounts = it.getBoolean(KEY_SWITCHING_ACCOUNTS, false)
        }

        setContentView(R.layout.activity_navigation_base)
        header = HeaderViews()

        header.boxAccountsSwitch.setOnClickListener { isSwitchingAccounts = !isSwitchingAccounts }
        header.butAccountEdit.apply {
            setImageDrawable(
                    getIconics(GoogleMaterial.Icon.gmd_edit)
                            .actionBar()
            )
            setOnClickListener { onEditAccountButtonClick(it) }
        }

        setNavigationViewAppIconDrawable(AndroidUtils.getActivityIcon(this, componentName))
        setHeaderBackgroundColor(AndroidUtils.getColorFromAttr(this,
                R.attr.colorPrimaryDark, Color.BLACK))

        header.txtAppName.text = AndroidUtils.getAppLabel(this)

        setSupportActionBar(toolbar)

        drawerToggle = ActionBarDrawerToggle(this, drawer, toolbar,
                R.string.navigation_drawer_open, R.string.navigation_drawer_close)
                .also {
                    drawer.addDrawerListener(it)
                    it.syncState()
                }

        navigationView.setNavigationItemSelectedListener listener@ {
            drawer.closeDrawer(GravityCompat.START)

            if (enableSwitchingAccounts && switchingAccounts) {
                isSwitchingAccounts = false
                return@listener onAccountNavigationItemSelected(it)
            }

            return@listener onNavigationItemSelected(it)
        }
        invalidateNavigationMenu()
    }

    override fun onPostCreate(savedInstanceState: Bundle?, persistentState: PersistableBundle?) {
        super.onPostCreate(savedInstanceState, persistentState)
        drawerToggle.syncState()
    }

    fun setNavigationViewAppIconBitmap(bm: Bitmap) =
            header.imgAppIcon.setImageBitmap(bm)

    fun setNavigationViewAppIconDrawable(drawable: Drawable?) =
            header.imgAppIcon.setImageDrawable(drawable)

    fun setNavigationViewAppIconResource(@DrawableRes resId: Int) =
            header.imgAppIcon.setImageResource(resId)

    fun setHeaderBackground(drawable: Drawable) =
            ViewCompat.setBackground(header.boxHeader, drawable)

    fun setHeaderBackgroundColor(@ColorInt color: Int) =
            header.boxHeader.setBackgroundColor(color)

    fun setHeaderBackgroundResource(@DrawableRes resId: Int) =
            header.boxHeader.setBackgroundResource(resId)

    fun setNavigationTitle(title: CharSequence) {
        header.txtAppName.text = title
    }

    override fun onSaveInstanceState(outState: Bundle, outPersistentState: PersistableBundle) {
        super.onSaveInstanceState(outState, outPersistentState)
        outState.putBoolean(KEY_SWITCHING_ACCOUNTS, switchingAccounts)
    }

    override fun onBackPressed() {
        drawer.takeIf { it.isDrawerOpen(GravityCompat.START) }?.let {
            it.closeDrawer(GravityCompat.START)
            return
        }

        // If class of current fragment differs from main fragment class,
        // replace current fragment with main fragment.
        (currentFragment to mainFragmentClass)
                .takeIf { it.first?.javaClass != it.second }
                ?.run {
                    // If main fragment class is not null, replace current fragment
                    // with main fragment, if it's null remove current fragment
                    second?.apply { replaceFragment(this) }
                            ?: removeCurrentFragment()
                    return
                }

        super.onBackPressed()
    }

    @SuppressLint("RestrictedApi")
    fun invalidateNavigationMenu() {
        if (enableSwitchingAccounts) {
            header.txtAccountName.text = onUpdateActiveAccountName()
            header.imgAccountsSwitchArrow.setImageDrawable(
                    getIconics(
                            if (switchingAccounts) GoogleMaterial.Icon.gmd_arrow_drop_up
                            else GoogleMaterial.Icon.gmd_arrow_drop_down
                    ).actionBar()
            )

            header.boxAppName.visibility = View.GONE
            header.boxAccountsSwitch.visibility = View.VISIBLE
        } else {
            header.boxAccountsSwitch.visibility = View.GONE
            header.boxAppName.visibility = View.VISIBLE
        }

        // Extract menu presenter from navigation view. This allows us to control navigation menu updating.
        val presenter: NavigationMenuPresenter? = try {
            NavigationView::class.java.getDeclaredField("mPresenter")?.apply {
                isAccessible = true
            }?.let {
                // Yeah, you are right, next line is really big and dirty HACK.
                it.get(navigationView) as NavigationMenuPresenter?
            }
        } catch (e: Exception) {
            Log.w(LOG_TAG, "invalidateNavigationMenu() -> " +
                    "Cannot extract menu presenter form navigation view: " +
                    "Failed to get menu presenter field from navigation view instance -> " +
                    "Falling back to updating navigation menu after every change", e)
            null
        }

        try {
            presenter?.setUpdateSuspended(true)

            navigationView.menu.apply {
                clear()

                if (enableSwitchingAccounts && switchingAccounts) {
                    onCreateAccountNavigationMenu(this)
                    // setupMenuItemsClickListeners(this);
                } else {
                    onCreateNavigationMenu(this)
                    // setupMenuItemsClickListeners(this);
                }
            }

            resetNavigationView(currentFragment)
        } finally {
            presenter?.run {
                setUpdateSuspended(false)
                updateMenuView(false)
            }
        }
    }

    private fun resetNavigationView(currentFragment: Fragment?) {
        if (enableSwitchingAccounts && switchingAccounts) {
            onUpdateSelectedAccountNavigationMenuItem(currentFragment, navigationView.menu)
            header.txtAccountName.text = onUpdateActiveAccountName()
        } else {
            onUpdateSelectedNavigationMenuItem(currentFragment, navigationView.menu)
        }
    }

    protected open fun onCreateNavigationMenu(menu: Menu): Boolean = false

    protected open fun onCreateAccountNavigationMenu(menu: Menu): Boolean = false

    /**
     * Override this method to enable navigation menu items selecting
     *
     * @param currentFragment current fragment
     * @param menu            navigation menu
     * @return true if checked state of any item was changed
     */
    protected open fun onUpdateSelectedNavigationMenuItem(currentFragment: Fragment?, menu: Menu): Boolean {
        if (currentFragment != null) {
            Log.e(LOG_TAG, "onUpdateSelectedNavigationMenuItem(currentFragment=$currentFragment, menu=$currentFragment) -> " +
                    "Cannot detect selected item for ${currentFragment.javaClass}, override method " +
                    "onUpdateSelectedNavigationMenuItem() and implement your own selected item detection.")
            return false
        }


        for (i in 0 until menu.size()) {
            menu.getItem(i).isChecked = false
        }
        return true
    }

    protected open fun onUpdateSelectedAccountNavigationMenuItem(currentFragment: Fragment?, menu: Menu): Boolean {
        Log.e(LOG_TAG, "onUpdateSelectedAccountNavigationMenuItem(currentFragment=$currentFragment, menu=$currentFragment) -> " +
                "Cannot update selected account item, override method onUpdateSelectedAccountNavigationMenuItem() " +
                "and implement your own selected item detection.")
        return false
    }

    open fun onNavigationItemSelected(item: MenuItem): Boolean = false

    open fun onAccountNavigationItemSelected(item: MenuItem): Boolean = false

    protected open fun onUpdateActiveAccountName(): CharSequence {
        Log.e(LOG_TAG, "onUpdateActiveAccountName() -> Cannot get active account name, " +
                "override method onUpdateActiveAccountName() and implement active account name detecting.")
        return ""
    }

    protected open fun onEditAccountButtonClick(v: View): Boolean = false

    override fun onAttachFragment(fragment: Fragment) {
        super.onAttachFragment(fragment)

        // Check if navigation view is available and if fragment is current fragment and if so, reset navigation view.
        if (navigationView != null && BaseFragmentActivity.FRAGMENT_TAG_CURRENT == fragment.tag)
            resetNavigationView(fragment)
    }

    @SuppressLint("PrivateResource")
    override fun onBeforeReplaceFragment(ft: FragmentTransaction, fragment: Fragment?) {
        ft.setCustomAnimations(R.anim.abc_fade_in, R.anim.abc_fade_out)
    }
}
